src/node_api.cc

```diff
@@ -444,6 +437,120 @@ class ThreadSafeFunction : public node::AsyncResource {
+  }
+
+  inline double get_async_id() const { return async_id_; }
+
+  inline double get_trigger_async_id() const { return trigger_async_id_; }
+
+  v8::MaybeLocal<v8::Value> MakeCallback(const v8::Local<v8::Object> recv,
+                                         const v8::Local<v8::Function> cb,
+                                         int argc,
+                                         v8::Local<v8::Value>* argv) {
+    if (*TRACE_EVENT_API_GET_CATEGORY_GROUP_ENABLED(
+            TRACING_CATEGORY_NODE1(async_hooks))) {
+      TRACE_EVENT_NESTABLE_ASYNC_BEGIN0(TRACING_CATEGORY_NODE1(async_hooks),
+                                        "napi_async_context_callback",
+                                        static_cast<int64_t>(get_async_id()));
+    }
+
+    node::async_context context{get_async_id(), get_trigger_async_id()};
+    v8::MaybeLocal<v8::Value> ret = node::InternalMakeCallback(
+        node_env(), resource(), recv, cb, argc, argv, context);
+
+    if (*TRACE_EVENT_API_GET_CATEGORY_GROUP_ENABLED(
+            TRACING_CATEGORY_NODE1(async_hooks))) {
+      TRACE_EVENT_NESTABLE_ASYNC_END0(TRACING_CATEGORY_NODE1(async_hooks),
+                                      "napi_async_context_callback",
+                                      static_cast<int64_t>(get_async_id()));
+    }
+    return ret;
+  }
+
+ private:
+  inline void EmitAsyncInit(v8::Local<v8::Object> resource,
+                            v8::Local<v8::String> resource_name) {
+    if (*TRACE_EVENT_API_GET_CATEGORY_GROUP_ENABLED(
+            TRACING_CATEGORY_NODE1(async_hooks))) {
+      auto data = node::tracing::TracedValue::Create();
+      node::Utf8Value resource_name_utf8(node_env()->isolate(), resource_name);
+      data->SetInteger("executionAsyncId",
+                       static_cast<int64_t>(node_env()->execution_async_id()));
+      data->SetInteger("triggerAsyncId",
+                       static_cast<int64_t>(get_trigger_async_id()));
+      TRACE_EVENT_NESTABLE_ASYNC_BEGIN1(TRACING_CATEGORY_NODE1(async_hooks),
+                                        "napi_async_context",
+                                        static_cast<int64_t>(get_async_id()),
+                                        "data",
+                                        std::move(data));
+    }
+
+    node::AsyncWrap::EmitAsyncInit(
+        node_env(), resource, resource_name, async_id_, trigger_async_id_);
+  }
+
+  inline void EmitAsyncDestroy() {
+    if (*TRACE_EVENT_API_GET_CATEGORY_GROUP_ENABLED(
+            TRACING_CATEGORY_NODE1(async_hooks))) {
+      TRACE_EVENT_NESTABLE_ASYNC_END0(TRACING_CATEGORY_NODE1(async_hooks),
+                                      "napi_async_context",
+                                      static_cast<int64_t>(get_async_id()));
+    }
+    node::AsyncWrap::EmitDestroy(node_env(), async_id_);
+    resource_.Reset();
+  }
+
+  node_napi_env env_;
+  double async_id_;
+  double trigger_async_id_;
+  v8::Global<v8::Object> resource_;
+};
```

Comment:
This seems to re-implement `AsyncResource` from node.h almost 1:1 in terms of functionality, maybe we could just use that class instead?